subprojects/core/src/main/java/org/gradle/api/internal/provider/DefaultProviderFactory.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.file.ConfigurableFileTree;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.FileTree;
+import org.gradle.api.internal.file.FileOperations;
+import org.gradle.api.provider.PropertyState;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
+
+import java.util.Collections;
+import java.util.concurrent.Callable;
+
+public class DefaultProviderFactory implements ProviderFactory {
+
+    private final FileOperations fileOperations;
+
+    public DefaultProviderFactory(FileOperations fileOperations) {
+        this.fileOperations = fileOperations;
+    }
+
+    public <T> Provider<T> provider(final Callable<T> value) {
+        if (value == null) {
+            throw new InvalidUserDataException("Value cannot be null");
+        }
+
+        return new DefaultProvider(value);
+    }
+
+    @Override
+    public <T> PropertyState<T> property(Class<T> clazz) {
+        if (clazz == null) {
+            throw new InvalidUserDataException("Class cannot be null");
+        }
+
+        PropertyState<T> propertyState = new DefaultPropertyState<T>();
+
+        if (clazz == Boolean.class) {
+            ((PropertyState<Boolean>) propertyState).set(Boolean.FALSE);
+        } else if (clazz == Byte.class) {
+            ((PropertyState<Byte>) propertyState).set(Byte.valueOf((byte) 0));
+        } else if (clazz == Short.class) {
+            ((PropertyState<Short>) propertyState).set(Short.valueOf((short) 0));
+        } else if (clazz == Integer.class) {
+            ((PropertyState<Integer>) propertyState).set(Integer.valueOf(0));
+        } else if (clazz == Long.class) {
+            ((PropertyState<Long>) propertyState).set(Long.valueOf(0));
+        } else if (clazz == Float.class) {
+            ((PropertyState<Float>) propertyState).set(Float.valueOf(0));
+        } else if (clazz == Double.class) {
+            ((PropertyState<Double>) propertyState).set(Double.valueOf(0));
+        } else if (clazz == Character.class) {
+            ((PropertyState<Character>) propertyState).set(new Character('\0'));
+        } else if (clazz == FileCollection.class || clazz == ConfigurableFileCollection.class) {
```

Comment:
We don't want people to use `Provider<FileCollection>`, we want them to use `ConfigurableFileCollection` or `Provider<Set<File>>` instead. This is because `FileCollection` already is a provider for a set of files, even though it doesn't yet extend `Provider<Set<File>>` (it's not a set of files, it's something that provides a set of files lazily. which is our definition of what a provider is). It doesn't make sense to have a provider of a provider of T.

So, for now we should leave these out I think. Same for `Provider<FileTree>`.